odxtools/ecu_variant_matching.py

```diff
@@ -39,40 +39,48 @@ class State(Enum):
         MATCH = 2
 
     @staticmethod
-    def get_ident_service(diag_layer: DiagLayer, matching_param: MatchingParameter) -> DiagService:
+    def get_ident_service(
+        diag_layer: DiagLayer, matching_param: MatchingParameter
+    ) -> DiagService:
         service_name = matching_param.diag_comm_snref
         # TODO this is not working since NamedItemList.__contains__() is not implemented
-        #assert service_name in diag_layer.services
+        # assert service_name in diag_layer.services
         service = diag_layer.services[service_name]
         assert isinstance(service, DiagService)
         return service
 
     @staticmethod
-    def encode_ident_request(diag_layer: DiagLayer, matching_param: MatchingParameter) -> bytearray:
+    def encode_ident_request(
+        diag_layer: DiagLayer, matching_param: MatchingParameter
+    ) -> bytes:
         service = EcuVariantMatcher.get_ident_service(diag_layer, matching_param)
-        return service.encode_request()
+        return bytes(service.encode_request())
 
     @staticmethod
-    def decode_ident_response(diag_layer: DiagLayer, matching_param: MatchingParameter, response: bytearray) -> str:
+    def decode_ident_response(
+        diag_layer: DiagLayer,
+        matching_param: MatchingParameter,
+        response: Union[bytes, bytearray],
+    ) -> str:
         service = EcuVariantMatcher.get_ident_service(diag_layer, matching_param)
         assert service.positive_responses is not None
         resp_decoded = service.positive_responses[0].decode(response)
         assert matching_param.out_param_if_snref in resp_decoded
         return resp_decoded[matching_param.out_param_if_snref]
 
     def __init__(self, ecu_variant_candidates: List[DiagLayer], use_cache: bool = True):
-        
+
         self.ecus = ecu_variant_candidates
         for ecu in self.ecus:
             assert ecu.variant_type == DIAG_LAYER_TYPE.ECU_VARIANT
 
         self.use_cache = use_cache
-        self.req_resp_cache: Dict[ByteString, Union[bytes, bytearray]] = {}
-        self._recent_ident_response: Optional[Union[bytes, bytearray]] = None
+        self.req_resp_cache: Dict[ByteString, bytes] = {}
+        self._recent_ident_response: Optional[bytes] = None
 
         self._state = EcuVariantMatcher.State.PENDING
 
-    def request_loop(self):
+    def request_loop(self) -> Generator[bytes, None, None]:
```

Comment:
do really you want to stick to this name for this generator? IMO it is not a loop, so `request_proceedure()` or `request_sequence()` would be more fitting, but if you insist, I'm okay with it..

Reply:
The term loop is inspired from control theory rather than computer science. I like it because it contains the notion of feedback, which is happening inside.